### PR TITLE
Expose generic update parameters on existing custom updates

### PR DESCRIPTION
**Pull request by tjprescott**
This PR exposes the existing generic update arguments (add, set, remove) on most of the existing custom update commands. This preserves the usability of the custom command while giving the power and flexibility of the generic update. There are some specify "update" commands that are not wrapped in this PR because I'd like to confer with @yugangw-msft on these commands:
- ad app update 
- resource policy update
- vm nic update

### CODE REVIEW

File: src/azure/cli/commands/arm.py [hunk 1]
```diff
@@ -226,6 +252,11 @@ def __call__(self, parser, namespace, values, option_string=None):
     cmd = CliCommand(name, handler, simple_output_query=simple_output_query)
     cmd.arguments.update(set_arguments)
     cmd.arguments.update(get_arguments)
+    if function_arguments:
+        cmd.arguments.update(function_arguments)
+    cmd.arguments.pop('instance', None) # inherited from custom_function(instance, ...)
+    cmd.arguments.pop('parent', None)
+    cmd.arguments.pop('expand', None) # possibly inherited from the getter
```
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: >cmd.arguments.pop('expand', None) # possibly inherited from the getter [](start = 4, length = 70)

I'm hesitant to pop an arg we don't own, it seems like this is also a corner case (not many commands using this).  Can we just hide this on the individual commands that have them?

File: src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py [hunk 2]
```diff
@@ -300,7 +303,7 @@ def body(self):
 class NetworkLoadBalancerSubresourceScenarioTest(ResourceGroupVCRTestBase):
 
     def __init__(self, test_method):
-        super(NetworkLoadBalancerSubresourceScenarioTest, self).__init__(__file__, test_method)
+        super(NetworkLoadBalancerSubresourceScenarioTest, self).__init__(__file__, test_method, debug=True)
```
Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py]: >, debug=True [](start = 94, length = 12)

this should be removed

Comment (BurtBiel) [inline src/command_modules/azure-cli-network/azure/cli/command_modules/network/tests/test_network_commands.py]: these tend to sneak through easily, it almost might make sense to set them via environment variable :)

---
In reply to: [75965252](https://github.com/Azure/azure-cli/pull/735#discussion_r75965252) [](ancestors = 75965252)

File: src/azure/cli/commands/arm.py [hunk 3]
```diff
@@ -160,11 +160,22 @@ def required_values_validator(namespace):
...
+def _get_child_property(parent, collection_name, item_name):
+    items = getattr(parent, collection_name)
+    result = next((x for x in items if x.name.lower() == item_name.lower()), None)
+    if not result:
+        raise CLIError("Property '{}' does not exist".format(item_name))
+    else:
+        return result
+
+def cli_generic_update_command(name, getter, setter, factory=None, setter_arg_name='parameters', #pylint:disable=too-many-arguments
+                               simple_output_query=None, child_arg_type=None,
```
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: >child_arg_type [](start = 57, length = 14)

we should rename this to something like child_collection_name

File: src/azure/cli/commands/arm.py [hunk 4]
```diff
@@ -160,11 +160,22 @@ def required_values_validator(namespace):
 
 APPLICATION.register(APPLICATION.COMMAND_TABLE_LOADED, add_id_parameters)
 
-def register_generic_update(name, getter, setter, factory=None, setter_arg_name='parameters', #pylint:disable=too-many-arguments
-                            simple_output_query=None, custom_handlers=None):
+def _get_child_property(parent, collection_name, item_name):
+    items = getattr(parent, collection_name)
+    result = next((x for x in items if x.name.lower() == item_name.lower()), None)
```
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: >x.name [](start = 39, length = 6)

are there any cases where there is not an x.name?  Perhaps the SDK has a .vm_name, etc.

File: src/azure/cli/commands/arm.py [hunk 5]
```diff
@@ -160,11 +160,22 @@ def required_values_validator(namespace):
 
 APPLICATION.register(APPLICATION.COMMAND_TABLE_LOADED, add_id_parameters)
 
-def register_generic_update(name, getter, setter, factory=None, setter_arg_name='parameters', #pylint:disable=too-many-arguments
-                            simple_output_query=None, custom_handlers=None):
+def _get_child_property(parent, collection_name, item_name):
```
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: >_get_child_property [](start = 4, length = 19)

this would probably be better renamed to something like _get_item since we're retrieving objects instead of properties

File: src/azure/cli/commands/arm.py [hunk 6]
```diff
@@ -176,16 +187,29 @@ def handler(args):
...
+        if child_arg_type:
+            parent = getter(client, **getterargs) if client else getter(**getterargs)
+            instance = _get_child_property(parent, child_arg_type, args.get(child_arg_name))
+        else:
+            parent = None
+            instance = getter(client, **getterargs) if client else getter(**getterargs)
+
+        # pass instance to the custom_function, if provided
+        if custom_function:
+            custom_func_args = {k: v for k, v in args.items() if k in function_arguments.keys()}
```
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: >.keys() [](start = 88, length = 7)

might not need this

File: src/azure/cli/commands/arm.py (hunk 1 above)
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: Yes, but then you have multiple ignore entries in the params file for something you can take care of here. Is there any case where you would want the getter part of generic update to retain the expand parameter?

---
In reply to: [75964122](https://github.com/Azure/azure-cli/pull/735#discussion_r75964122) [](ancestors = 75964122)

File: src/azure/cli/commands/arm.py (hunk 4 above)
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: I haven't run into one yet.

---
In reply to: [75969337](https://github.com/Azure/azure-cli/pull/735#discussion_r75969337) [](ancestors = 75969337)

File: src/azure/cli/commands/arm.py (hunk 1 above)
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: Aren't there only about 5 commands that have expand?  If so, I'd vote for manually hiding them

---
In reply to: [76079251](https://github.com/Azure/azure-cli/pull/735#discussion_r76079251) [](ancestors = 76079251,75964122)

File: src/azure/cli/commands/arm.py (hunk 4 above)
Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: maybe we should change this to if getattr(x, 'name', '').lower() so we fail with a meaningful error message

---
In reply to: [76079355](https://github.com/Azure/azure-cli/pull/735#discussion_r76079355) [](ancestors = 76079355,75969337)

File: src/azure/cli/commands/arm.py (hunk 3 above)
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: I went with 'child_collection_prop_name' since it's the property name of the child collection on the client.

---
In reply to: [75968958](https://github.com/Azure/azure-cli/pull/735#discussion_r75968958) [](ancestors = 75968958)

File: src/azure/cli/commands/arm.py (hunk 4 above)
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: I changed it to be like the 'setter_params_arg_name' so we can override it.

---
In reply to: [76099476](https://github.com/Azure/azure-cli/pull/735#discussion_r76099476) [](ancestors = 76099476,76079355,75969337)

File: src/azure/cli/commands/arm.py (hunk 1 above)
Comment (tjprescott) [inline src/azure/cli/commands/arm.py]: They accomplish the same thing, though preventing the argument from being added to the argument space to begin with is a better approach than adding it and trying to suppress it. Isn't this why you did the same thing for the setter's parameters arg?

---
In reply to: [76099240](https://github.com/Azure/azure-cli/pull/735#discussion_r76099240) [](ancestors = 76099240,76079251,75964122)

Comment (BurtBiel) [inline src/azure/cli/commands/arm.py]: The setters were always present, so popping them followed a standard pattern.  However, if we view this as a "blacklist" like we have for ignoring built-in sdk params, I can see how they are similar.

---
In reply to: [76142981](https://github.com/Azure/azure-cli/pull/735#discussion_r76142981) [](ancestors = 76142981,76099240,76079251,75964122)

**Comment by BurtBiel** [comment]
:shipit:
